plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -107,7 +107,8 @@ unroll = runQuote $ do
 -- | 'fix' as a PLC term.
 --
 -- > /\(a b :: *) -> \(f : (a -> b) -> a -> b) ->
--- >    unroll {a -> b} (iwrap selfF (a -> b) \(s : self (a -> b)) \(x : a) -> f (unroll {a -> b} s) x)
+-- >     unroll {a -> b} (iwrap selfF (a -> b) \(s : self (a -> b)) ->
+-- >         f (\(x : a) -> unroll {a -> b} s x))
```

Comment:
Is `f (\(x : a) -> unroll {a -> b} s x))` the same as `f (unroll {a -> b} s))`?

Hmm I wonder why we defined `fix` that way before. Isn't this the more obvious definition? This is exactly the z combinator.

Reply:
> Is `f (\(x : a) -> unroll {a -> b} s x))` the same as `f (unroll {a -> b} s))`?

In a lazy language, yes. In a strict language that transformations turns the Z combinator into the Y combinator and the latter diverges.

> Hmm I wonder why we defined `fix` that way before. Isn't this the more obvious definition? This is exactly the z combinator.

I don't know! It was me who picked the old version and I have no idea why. Maybe because it's lazier and I wanted to avoid doing unnecessary work or something. Maybe it was completely accidental. I just read through all our discussions at the time and we always used the new version as the default. Somehow the old one still got merged. A mystery. Quite likely just a stupid mistake on my part.

Reply:
> In a lazy language, yes. In a strict language that transformation turns the Z combinator into the Y combinator and the latter diverges.

Right, of course!